src/line_buffer.rs

```diff
@@ -194,13 +212,280 @@ impl LineBuffer {
+        }
+    }
 }
 
 /// Match any sequence of characters that are considered a word boundary
 fn is_word_boundary(s: &str) -> bool {
     !s.chars().any(char::is_alphanumeric)
 }
 
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn buffer_with(content: &str) -> LineBuffer {
+        let mut line_buffer = LineBuffer::new();
+        line_buffer.insert_str(content);
+
+        line_buffer
+    }
+
+    #[test]
+    fn test_new_buffer_is_empty() {
+        let line_buffer = LineBuffer::new();
+        assert!(line_buffer.is_empty())
+    }
+
+    #[test]
+    fn insert_str_updates_insertion_point_point_correctly() {
+        let mut line_buffer = LineBuffer::new();
+        line_buffer.insert_str("this is a command");
+
+        let expected_updated_insertion_point = InsertionPoint {
+            line: 0,
+            offset: 17,
+        };
+
+        assert_eq!(
+            expected_updated_insertion_point,
+            line_buffer.insertion_point()
+        );
+    }
+
+    #[test]
+    fn insert_char_updates_insertion_point_point_correctly() {
+        let mut line_buffer = LineBuffer::new();
+        line_buffer.insert_char('c');
+
+        let expected_updated_insertion_point = InsertionPoint { line: 0, offset: 1 };
+
+        assert_eq!(
+            expected_updated_insertion_point,
+            line_buffer.insertion_point()
+        );
+    }
+
+    #[test]
+    fn delete_left_grapheme_works() {
+        let mut line_buffer = buffer_with("This is a test");
+        line_buffer.delete_left_grapheme();
+
+        let expected_line_buffer = buffer_with("This is a tes");
+
+        assert_eq!(expected_line_buffer, line_buffer);
+    }
+
+    #[test]
+    fn delete_left_grapheme_works_with_emojis() {
+        let mut line_buffer = buffer_with("This is a test 😊");
+        line_buffer.delete_left_grapheme();
+
+        let expected_line_buffer = buffer_with("This is a test ");
+
+        assert_eq!(expected_line_buffer, line_buffer);
+    }
+
+    #[test]
+    fn delete_left_grapheme_on_an_empty_buffer_is_a_no_op() {
+        let mut line_buffer = buffer_with("");
+        line_buffer.delete_left_grapheme();
+
+        let expected_line_buffer = buffer_with("");
+
+        assert_eq!(expected_line_buffer, line_buffer);
+    }
+
+    #[test]
+    fn delete_right_grapheme_works() {
+        let mut line_buffer = buffer_with("This is a test");
+        line_buffer.move_left();
+        line_buffer.delete_right_grapheme();
+
+        let expected_line_buffer = buffer_with("This is a tes");
+
+        assert_eq!(expected_line_buffer, line_buffer);
+    }
+
+    #[test]
+    fn delete_right_grapheme_works_with_emojis() {
+        let mut line_buffer = buffer_with("This is a test 😊");
+        line_buffer.move_left();
+        line_buffer.delete_right_grapheme();
+
+        let expected_line_buffer = buffer_with("This is a test ");
+
+        assert_eq!(expected_line_buffer, line_buffer);
+    }
```

Comment:
In python we have things like `pytest.paramertize` which can make this sort of testing super easy. I am not sure what is the idomatic way to do this in rust. I was looking at libs like `rstest`,`quickcheck`,`proptest` etc that can help in this regard but they seem either too limited or too complex. Let me know your thoughts?

Reply:
I'm not aware of an established fixture-based testing framework. I tend to formulate my tests for a whole aspect and have sometimes more than one assert in my rust tests if it makes sense. Other than that, functions in the test module are fair game.

Reply:
For now, I will keep these as-is. I will update if folks decide that we want to include any of these.